File: pkg/geometry/cone.go
package geometry

import (
	"math"

	"sif/gogs/eric/relray/pkg/vec"
)

// Cone is a capped cone along the local Y axis.
// Base at Y=0 with the given Radius, apex at Y=Height with radius 0.
// Use a Transformed wrapper to position and orient it.
type Cone struct {
	Radius float64
	Height float64
}

func (c *Cone) Intersect(origin, dir vec.Vec3, tMin, tMax float64) (Hit, bool) {
	// Cone equation: x² + z² = (r/h)² * (h - y)²
	// Let k = r/h
	k := c.Radius / c.Height
	k2 := k * k

	// Substitute ray: p = o + t*d
	// (ox+t*dx)² + (oz+t*dz)² = k²*(h - oy - t*dy)²
	hy := c.Height - origin.Y
	a := dir.X*dir.X + dir.Z*dir.Z - k2*dir.Y*dir.Y
	b := 2*(origin.X*dir.X+origin.Z*dir.Z) + 2*k2*hy*dir.Y
	cc := origin.X*origin.X + origin.Z*origin.Z - k2*hy*hy

	var bestT float64
	var bestNormal vec.Vec3
	found := false

	disc := b*b - 4*a*cc
	if disc >= 0 && math.Abs(a) > 1e-12 {
		sqrtDisc := math.Sqrt(disc)
		for _, t := range [2]float64{(-b - sqrtDisc) / (2 * a), (-b + sqrtDisc) / (2 * a)} {
			if t < tMin || t > tMax {
				continue
			}
			p := origin.Add(dir.Scale(t))
			if p.Y >= 0 && p.Y <= c.Height {
				if !found || t < bestT {
					bestT = t
					// Cone normal: for point (x, y, z) on cone surface,
					// outward normal has radial component and downward Y component
					r := math.Sqrt(p.X*p.X + p.Z*p.Z)
					if r > 1e-12 {
						bestNormal = vec.Vec3{
							X: p.X / r,
							Y: k,
							Z: p.Z / r,
						}.Normalize()
					} else {
						bestNormal = vec.Vec3{Y: 1} // at apex
					}
					found = true
				}
				break
			}
		}
	}

	// Test base cap (Y=0)
	if math.Abs(dir.Y) > 1e-12 {
		t := -origin.Y / dir.Y
		if t >= tMin && t <= tMax {
			p := origin.Add(dir.Scale(t))
			if p.X*p.X+p.Z*p.Z <= c.Radius*c.Radius {
				if !found || t < bestT {
					bestT = t
					bestNormal = vec.Vec3{Y: -1}
					found = true
				}
			}
		}
	}

	if !found {
		return Hit{}, false
	}
	p := origin.Add(dir.Scale(bestT))
	h := Hit{T: bestT, Point: p}
	h.SetFaceNormal(dir, bestNormal)
	return h, true
}

File: rust/src/geometry/cone.rs
use crate::vec::Vec3;
use super::{Hit, Shape};

pub struct Cone {
    pub radius: f64,
    pub height: f64,
}

impl Shape for Cone {
    fn intersect(&self, origin: Vec3, dir: Vec3, t_min: f64, t_max: f64) -> Option<Hit> {
        let k = self.radius / self.height;
        let k2 = k * k;

        let hy = self.height - origin.y;
        let a = dir.x * dir.x + dir.z * dir.z - k2 * dir.y * dir.y;
        let b = 2.0 * (origin.x * dir.x + origin.z * dir.z) + 2.0 * k2 * hy * dir.y;
        let cc = origin.x * origin.x + origin.z * origin.z - k2 * hy * hy;

        let mut best_t: f64 = 0.0;
        let mut best_normal = Vec3::default();
        let mut found = false;

        let disc = b * b - 4.0 * a * cc;
        if disc >= 0.0 && a.abs() > 1e-12 {
            let sqrt_disc = disc.sqrt();
            for t in [(-b - sqrt_disc) / (2.0 * a), (-b + sqrt_disc) / (2.0 * a)] {
                if t < t_min || t > t_max {
                    continue;
                }
                let p = origin + dir * t;
                if p.y >= 0.0 && p.y <= self.height {
                    if !found || t < best_t {
                        best_t = t;
                        let r = (p.x * p.x + p.z * p.z).sqrt();
                        best_normal = if r > 1e-12 {
                            Vec3 {
                                x: p.x / r,
                                y: k,
                                z: p.z / r,
                            }
                            .normalize()
                        } else {
                            Vec3 { x: 0.0, y: 1.0, z: 0.0 }
                        };
                        found = true;
                    }
                    break;
                }
            }
        }

        // Test base cap (Y=0)
        if dir.y.abs() > 1e-12 {
            let t = -origin.y / dir.y;
            if t >= t_min && t <= t_max {
                let p = origin + dir * t;
                if p.x * p.x + p.z * p.z <= self.radius * self.radius {
                    if !found || t < best_t {
                        best_t = t;
                        best_normal = Vec3 { x: 0.0, y: -1.0, z: 0.0 };
                        found = true;
                    }
                }
            }
        }

        if !found {
            return None;
        }
        let p = origin + dir * best_t;
        let mut h = Hit { t: best_t, point: p, ..Default::default() };
        h.set_face_normal(dir, best_normal);
        Some(h)
    }
}
